src/IssueViz.UnitTests/Editor/SelectedIssueTagging/Adornment/IssueLocationAdornmentTagger_CacheTests.cs

```diff
@@ -0,0 +1,177 @@
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            ThreadHelper.SetCurrentThreadAsUIThread();
+        }
+
+        [TestMethod]
+        public void CreateOrUpdate_DoesNotExist_ReturnsNewAdornment()
+        {
+            var testSubject = new CachingAdornmentFactory(ValidTextView);
+            var newLocViz1 = CreateLocationViz();
+            var newLocViz2 = CreateLocationViz();
+
+            // Create an adornment
+            var actual1 = testSubject.CreateOrUpdate(newLocViz1);
+
+            actual1.LocationViz.Should().Be(newLocViz1);
+            testSubject.CachedAdornments.Should().BeEquivalentTo(actual1);
+
+            // Create another adornment
+            var actual2 = testSubject.CreateOrUpdate(newLocViz2);
+
+            actual2.LocationViz.Should().Be(newLocViz2);
+            actual1.Should().NotBe(actual2);
+            testSubject.CachedAdornments.Should().BeEquivalentTo(actual1, actual2);
+        }
+
+        [TestMethod]
+        public void CreateOrUpdate_Exists_ReturnsExistingAdornment()
+        {
+            var testSubject = new CachingAdornmentFactory(ValidTextView);
+            var locViz1 = CreateLocationViz();
+            var locViz2 = CreateLocationViz();
+
+            // Populate the cache
+            var adornment1 = testSubject.CreateOrUpdate(locViz1);
+            var adornment2 = testSubject.CreateOrUpdate(locViz2);
+            testSubject.CachedAdornments.Should().BeEquivalentTo(adornment1, adornment2);
+
+            // Act
+            testSubject.CreateOrUpdate(locViz1).Should().Be(adornment1);
+            testSubject.CreateOrUpdate(locViz2).Should().Be(adornment2);
+
+            testSubject.CachedAdornments.Should().BeEquivalentTo(adornment1, adornment2);
+        }
+
+        [TestMethod]
+        public void CreateOrUpdate_Exists_AdornmentIsUpdated()
+        {
+            var lineSource1 = CreateFormattedLineSource(12d);
+            var lineSource2 = CreateFormattedLineSource(18d);
+            var textView = CreateWpfTextView(ValidSnapshot, lineSource1);
+            var testSubject = new CachingAdornmentFactory(textView);
+
+            var locViz = CreateLocationViz();
+
+            var actual1 = testSubject.CreateOrUpdate(locViz);
+            actual1.FontSize.Should().Be(12d); // sanity check
+
+            ChangedMockedLineSource(textView, lineSource2);
+
+            // Act
+            var actual2 = testSubject.CreateOrUpdate(locViz);
+            actual2.FontSize.Should().Be(18d);
+        }
+
+        [TestMethod]
+        public void RemoveUnused_EmptyCache_NoError()
+        {
+            var testSubject = new CachingAdornmentFactory(ValidTextView);
+
+            Action act = () => testSubject.RemoveUnused(new[] { ValidLocViz });
+            act.Should().NotThrow();
+        }
+
+        [TestMethod]
+        public void RemoveUnused_NoCurrentLocations_AllRemoved()
```

Comment:
Maybe
`Refresh_NoNewLocations_CacheCleared` ?

Reply:
"NoNewLocations" isn't very clear; the cache only needs to know the  entire set of current locations, not whether they are new or not.

I'll change it to `Refresh_NoCurrentLocations_CacheCleared`